lib/conjur/audit/webservice.rb

```diff
@@ -0,0 +1,71 @@
+require 'json'
+require 'uri'
+require 'sequel'
+
+# Simple Rack application to dump audit entries from a dataset,
+# filtered by the query string.
+# assumes schema:
+# (
+#     facility integer,
+#     severity integer,
+#     "timestamp" timestamp with time zone,
+#     hostname text,
+#     appname text,
+#     procid text,
+#     msgid text,
+#     sdata jsonb,
+#     message text
+# )
+
+# Note this is intentionally a separate rack app isolated from rest of Conjur.
+# This is to allow us to be flexible and spin it off into a separate gem and
+# have it run as a separate webservice, perhaps on a different host.
+
+module Conjur
+  module Audit
+    class Webservice
+      def initialize dataset
+        @messages = dataset
+
+        # ensure required extensions are loaded
+        dataset.db.extension :pg_json
+        Sequel.extension :pg_json_ops
+      end
+
+      def call env
+        filtered = filter messages, URI.decode_www_form(env['QUERY_STRING'])
+        if filtered.any?
+          result = filtered.all.to_json
+          [200, {'Content-Type' => 'text/json', 'Content-Length' => result.length}, [result]]
+        else
+          [404, {}, ["No records found"]]
+        end
+      end
+
+      # converts [a, b, c], d into {a => {b => {c => d}}}
+      def to_path_hash keys, value = nil
+        return to_path_hash [*keys, value] if value
+        return keys.first if keys.length == 1
+        { keys.first => to_path_hash(keys.drop 1) }
+      end
+
+      # filters dataset based on given query
+      # keys not in _fields_ are assumed to be in structured data
+      # eg. subject@43868/role=foo looks for [subject@43868 role="foo"]
+      def filter dataset, query
+        fields = %w(facility severity hostname appname procid msgid)
```

Comment:
It seems like we're going to want a higher level of abstraction here. Clients of this webservice are going to want to query in terms of Conjur objects, e.g. show me the events related to this resource. Using the database fields for filtering feels like it's leaking implementation detail.

Reply:
I both agree and not. Let me explain :)

It's not using 'database' fields (though indeed it translates the query directly into them, that fact alone being a -- not exposed -- implementation detail), but standard syslog fields defined by RFC, plus any structured data fields the entries may contain. As for the latter, RFC stipulates an authority responsible for specifying them (as registered with IANA), and so for Conjur-specific fields, we will need to publish a spec once they're settled (not only for using this service but also for external consumption eg. via syslog relay). In my view this makes it all a public interface, not an implementation detail.

That said, of course we'll want to add querying in terms of Conjur objects, but I feel we should add it on as a layer of smart sugar on top of this basic and generic functionality, which is meant to generically query structured syslog data, whether it's originally produced by Conjur or not (regardless of whether we will ever use it to query other events).

Reply:
OK, that description makes sense to me.

Where would you see the sugar layer being added, though? Anywhere outside this service seems like it would imply some amount of duplicated code (e.g. in each different client library).

Reply:
I'm not 100% sure and I'm open to suggestions. One idea I had is for Conjur-specific entries would be handled by `/audit/resource/*id` and `/audit/role/*id` or even `/audit/*id` and redirect or forward to this service with the specific query.

Although now that I think of it these couldn't be in fact implemented as simple wrappers for ONE query with this service, because it only supports conjunction of conditions, and a given role or resource can appear in many places in an audit entry. OTOH I wonder how important or useful it is to provide such general catch-all queries; yes, we had them in v4, and the fact that we ONLY had them is why the UI had to do so much filtering and transformation of the data, and that would probably be better-served by using more specific queries. Thoughts?

Reply:
Maybe you'll learn more by prototyping use of this service in the UI. We can go with this as-is for now, and update based on what you find.